word2gauss/embeddings.pyx

```diff
@@ -249,6 +249,87 @@ cdef class GaussianEmbedding:
         self.acc_grad_sigma = _acc_grad_sigma
         self.acc_grad_sigma_ptr = &_acc_grad_sigma[0]
 
+    def update(self, N, init_params={
+            'mu0': 0.1,
+            'sigma_mean0': 0.5,
+            'sigma_std0': 1.0
+        }):
+        '''
+        This function adds to the current np arrays: mu, sigma, acc_grad_mu and acc_grad_sigma to account for any new vocabulary that we need to introduce to the model for retraining.
+
+        If the original size of the vocabulary is N, and the new size is M
+        It adds new rows to self.mu to make it M x K
+        It increases the size of the other M x 1 dimensional arrays.
+
+        self = embeddings model
+        N = Total size of the vocabulary
+        init_params = initialization parameters to initialize new rows
+        '''
+        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] _mu
+        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] _sigma
+        cdef np.ndarray[DTYPE_t, ndim=1, mode='c'] _acc_grad_mu
+        cdef np.ndarray[DTYPE_t, ndim=1, mode='c'] _acc_grad_sigma
+        LOGGER.info("New vocab size %i" % N)
+        LOGGER.info("Old vocab size %i" % self.mu.shape[0])
+        if N > self.mu.shape[0]:
+
+            if self.covariance_type == SPHERICAL:
+                LOGGER.info("covariance is spherical")
+            else:
+                raise ValueError
+
+            # there are new words in the vocabulary
+            # add more rows
+            n_words = N - self.mu.shape[0]
+            LOGGER.info("%i new words" % n_words)
+
+            # update mu with n_words more rows initialized randomly
+            _mu = np.vstack([self.mu, init_params['mu0'] * np.ascontiguousarray(
+                np.random.randn(n_words, self.mu.shape[1]).astype(DTYPE))])
+            self.mu = _mu
+            LOGGER.info("mu updated with %i new rows" % n_words)
+
+            # update sigma
+            s = np.random.randn(n_words, 1).astype(DTYPE)
+            s *= init_params['sigma_std0']
+            s += init_params['sigma_mean0']
+            LOGGER.info("s created with %i new rows" % n_words)
+
+            _sigma = np.vstack([self.sigma, np.ascontiguousarray(
+                np.maximum(self.sigma_min, np.minimum(s, self.sigma_max)))])
+            LOGGER.info("sigma updated with %i new rows" % n_words)
+            self.sigma = _sigma
+
+            assert _mu.flags['C_CONTIGUOUS']
+            assert _sigma.flags['C_CONTIGUOUS']
```

Comment:
Can you also add some checks for the two `acc_grad_*`, they have similar requirements on data layout and structure.  And also add one for `'OWNDATA'`.  In most cases these aren't necessary but when they are then they are really useful.  It has saved me much heartache in the past in other projects.